pkg/output/output_test.go

```diff
@@ -0,0 +1,170 @@
+			name      string
+			expectErr string
+		}
+		writeActions []struct {
+			parent    string
+			name      string
+			contents  []string
+			expectErr string
+		}
+		expectedFiles []struct {
+			path     string
+			contents string
+		}
+	}{
+		{
+			name: "simple-operations",
+			dirActions: []struct {
+				parent    string
+				name      string
+				expectErr string
+			}{
+				{parent: "", name: "abc"},
+				{parent: "abc", name: "def"},
+				{parent: "abc/def", name: "ghi"},
+			},
+			writeActions: []struct {
+				parent    string
+				name      string
+				contents  []string
+				expectErr string
+			}{
+				{parent: "abc/def/ghi", name: "jkw", contents: []string{"file-contents-1", "file-contents-2"}},
+			},
+			expectedFiles: []struct {
+				path     string
+				contents string
+			}{
+				{"abc/def/ghi/jkw", "file-contents-1\nfile-contents-2\n"},
+			},
+		},
+		{
+			name: "non-existent directory creation",
+			dirActions: []struct {
+				parent    string
+				name      string
+				expectErr string
+			}{
+				{parent: "", name: "abc"},
+				{parent: "abc/def", name: "ghi", expectErr: "no such file or directory"},
+			},
+		},
+		{
+			name: "write invalid directory",
+			dirActions: []struct {
+				parent    string
+				name      string
+				expectErr string
+			}{
+				{
+					parent: "",
+					name:   "abc",
+				},
+			},
+			writeActions: []struct {
+				parent    string
+				name      string
+				contents  []string
+				expectErr string
+			}{
+				{
+					parent:    "def",
+					name:      "ghi",
+					contents:  []string{"test-content"},
+					expectErr: "no such file or directory",
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			tempdir, err := os.MkdirTemp(os.TempDir(), "test-*")
+			require.NoError(t, err)
+			defer os.RemoveAll(tempdir)
+			writer, err := NewFSWriter(tempdir)
+			require.NoError(t, err)
+			for _, a := range tc.dirActions {
+				err = writer.MakeDir(a.parent, a.name, 0700)
+				if a.expectErr != "" {
+					require.Error(t, err)
+					require.Contains(t, err.Error(), a.expectErr)
+				} else {
+					require.NoError(t, err)
+				}
+			}
+			for _, a := range tc.writeActions {
+				func() {
+					closeWriter, writer, err := writer.Writer(a.parent, a.name, 0700)
+					if a.expectErr == "" {
+						require.NoError(t, err)
+						defer func() {
+							require.NoError(t, closeWriter())
+						}()
+						for _, c := range a.contents {
+							_, err = writer.WriteString(fmt.Sprintf("%s\n", c))
+							require.NoError(t, err)
+						}
+					} else {
+						require.Error(t, err)
+						require.Contains(t, err.Error(), a.expectErr)
+					}
+				}()
+			}
+			for _, f := range tc.expectedFiles {
+				fPath := path.Join(tempdir, f.path)
+				contents, err := ioutil.ReadFile(fPath)
+				require.NoError(t, err)
+				require.Equal(t, f.contents, string(contents))
+			}
+		})
+	}
+}
+
+func TestFSWriterNonEmptyDir(t *testing.T) {
+	testDir, err := os.MkdirTemp(os.TempDir(), "test-dir-*")
```

Comment:
you can also use `ioutil.TempDir("", "test-dir")`. I think it would be also useful to add an indicator of your test name in the file, so when something fails you can later refer back to the file and debug if possible.

Please also try to clean the directories that you create (even when temp), I found myself deleting a couple of gigs the other day from my own project test files after a year of development.

Reply:
Error messages will have the full path of the failing directory.

Reply:
how come? they are never included in your require statements

Reply:
If any file operations fail then the error at the level will have the path of the file. But if we are cleaning up then all of this is moot anyway because any files in the failing tests will be deleted anyway.